pinot-plugins/pinot-stream-ingestion/pinot-pulsar/src/test/java/org/apache/pinot/plugin/stream/pulsar/PulsarConsumerTest.java

```diff
@@ -273,13 +269,31 @@ public void testPartitionLevelConsumerBatchMessages()
     }
   }
 
-  public MessageId getMessageIdForPartitionAndIndex(int partitionNum, int index) {
+  private static MessageBatch consumeMessageBatch(PartitionGroupConsumer consumer, MessageId startMsgId,
+      MessageId endMsgId, int expectedMsgCount)
+      throws TimeoutException {
+    int retryCount = 0;
+    MessageBatch messageBatch = null;
+    while (retryCount < DEFAULT_RETRY_COUNT) {
+      retryCount++;
+      messageBatch = consumer.fetchMessages(new MessageIdStreamOffset(startMsgId),
```

Comment:
do we know why this call may give fewer records? Is it a bug or something like messages not flushed entirely and available to consume or some messages are actually missing

Reply:
I dunno why this fixes the bug. My guess is that pulsar flush is async but I am not sure. 

GitHub Actions sometimes catch interesting bug b/c the machines are much less powerful

Reply:
I saw the publishRecordBatch was changed in #8017 - any chance @KKcorps can give some insights?

Reply:
This method was added to handle the scenarios where messages are published in batches which actually changes the format of MessageIds in Pulsar.

The only way to publish in batch is to enable async producer. In sync mode, all messages are flushed as soon as they are received.

Reply:
@xiangfu0 your suggestion also works. I can actually add a method like `waitForAllRecordsToFlush` and only then exit the setup.

Reply:
according to the documentation https://pulsar.apache.org/api/client/org/apache/pulsar/client/api/Producer.html 
- producer.flush() should wait until all the message finish publishing before returning (or throw exception); thus we can't check on the producer side
- only way to check and wait for all records reaches pulsar for consumption is to actually consume -- which is exactly what this test is doing. 
 
so I guess we should do is wrap the test with 
```
TestUtils.waitForCondition(aVoid -> { 
  // consume and assert 
}, timeoutMs, "")
```

sounds good?

Reply:
There is another approach - https://github.com/apache/pinot/pull/8554